File: 12/gemini-3.0-pro/go/main.go
package main

import (
"fmt"
"os"
"sort"
"strconv"
"strings"
)

type Point struct {
r, c int
}

type RegionTask struct {
w, h   int
counts []int
}

type AnchoredShape struct {
points []Point
}

func main() {
inputFile := "input.txt"
if len(os.Args) > 1 {
inputFile = os.Args[1]
}

data, err := os.ReadFile(inputFile)
if err != nil {
fmt.Printf("Input file %s not found.\n", inputFile)
return
}

content := strings.ReplaceAll(string(data), "\r\n", "\n")
parts := strings.Split(content, "\n\n")

shapes := make(map[int][][]Point)
var regions []RegionTask

for _, part := range parts {
lines := strings.Split(strings.TrimSpace(part), "\n")
if len(lines) == 0 {
continue
}

if strings.Contains(lines[0], ":") {
if strings.Contains(lines[0], "x") { // Region
for _, line := range lines {
if strings.TrimSpace(line) != "" {
parseRegion(line, &regions)
}
}
} else { // Shape
parseShape(lines, shapes)
}
}
}

solvedCount := 0
for _, region := range regions {
if solveRegion(region, shapes) {
solvedCount++
}
}

fmt.Printf("Day 12 Part 1: %d\n", solvedCount)
// fmt.Printf("Day 12 Part 2: 0\n")
}

func parseShape(lines []string, shapes map[int][][]Point) {
header := strings.TrimSuffix(strings.TrimSpace(lines[0]), ":")
id, err := strconv.Atoi(header)
if err != nil {
return
}

points := make(map[Point]bool)
for r := 1; r < len(lines); r++ {
for c := 0; c < len(lines[r]); c++ {
if lines[r][c] == '#' {
points[Point{r - 1, c}] = true
}
}
}
shapes[id] = generateVariations(points)
}

func parseRegion(line string, regions *[]RegionTask) {
p := strings.Split(line, ":")
dims := strings.Split(p[0], "x")
w, _ := strconv.Atoi(dims[0])
h, _ := strconv.Atoi(dims[1])

countStrs := strings.Fields(strings.TrimSpace(p[1]))
counts := make([]int, len(countStrs))
for i, s := range countStrs {
counts[i], _ = strconv.Atoi(s)
}
*regions = append(*regions, RegionTask{w, h, counts})
}

func generateVariations(original map[Point]bool) [][]Point {
variations := make([][]Point, 0)
seen := make(map[string]bool)

current := original
for i := 0; i < 2; i++ { // Flip
for j := 0; j < 4; j++ { // Rotate
normalized := normalize(current)
key := pointsToString(normalized)
if !seen[key] {
variations = append(variations, normalized)
seen[key] = true
}
current = rotate(current)
}
current = flip(original)
}
return variations
}

func rotate(points map[Point]bool) map[Point]bool {
newPoints := make(map[Point]bool)
for p := range points {
newPoints[Point{p.c, -p.r}] = true
}
return newPoints
}

func flip(points map[Point]bool) map[Point]bool {
newPoints := make(map[Point]bool)
for p := range points {
newPoints[Point{p.r, -p.c}] = true
}
return newPoints
}

func normalize(points map[Point]bool) []Point {
if len(points) == 0 {
return []Point{}
}
minR, minC := 1000000, 1000000
for p := range points {
if p.r < minR {
minR = p.r
}
if p.c < minC {
minC = p.c
}
}
res := make([]Point, 0, len(points))
for p := range points {
res = append(res, Point{p.r - minR, p.c - minC})
}
// Sort for consistent key generation
sort.Slice(res, func(i, j int) bool {
if res[i].r != res[j].r {
return res[i].r < res[j].r
}
return res[i].c < res[j].c
})
return res
}

func pointsToString(points []Point) string {
var sb strings.Builder
for _, p := range points {
sb.WriteString(fmt.Sprintf("%d,%d;", p.r, p.c))
}
return sb.String()
}

type SolverContext struct {
w, h           int
counts         map[int]int
types          []int
anchoredShapes map[int][]AnchoredShape
}

func solveRegion(region RegionTask, shapes map[int][][]Point) bool {
presents := make([]int, 0)
totalArea := 0
for i, count := range region.counts {
for j := 0; j < count; j++ {
presents = append(presents, i)
totalArea += len(shapes[i][0])
}
}

if totalArea > region.w*region.h {
return false
}

maxSkips := (region.w * region.h) - totalArea
grid := make([]bool, region.w*region.h)

presentTypesMap := make(map[int]bool)
for _, p := range presents {
presentTypesMap[p] = true
}
presentTypes := make([]int, 0, len(presentTypesMap))
for p := range presentTypesMap {
presentTypes = append(presentTypes, p)
}
sort.Slice(presentTypes, func(i, j int) bool {
return len(shapes[presentTypes[i]][0]) > len(shapes[presentTypes[j]][0])
})

counts := make(map[int]int)
for _, p := range presents {
counts[p]++
}

anchoredShapes := make(map[int][]AnchoredShape)
for _, typeID := range presentTypes {
anchoredShapes[typeID] = make([]AnchoredShape, 0)
for _, v := range shapes[typeID] {
// v is already sorted in normalize
anchor := v[0]
points := make([]Point, len(v))
for i, p := range v {
points[i] = Point{p.r - anchor.r, p.c - anchor.c}
}
anchoredShapes[typeID] = append(anchoredShapes[typeID], AnchoredShape{points})
}
}

ctx := SolverContext{
w:              region.w,
h:              region.h,
counts:         counts,
types:          presentTypes,
anchoredShapes: anchoredShapes,
}

return backtrack(0, grid, ctx, maxSkips)
}

func backtrack(idx int, grid []bool, ctx SolverContext, skipsLeft int) bool {
for idx < len(grid) && grid[idx] {
idx++
}

if idx == len(grid) {
for _, c := range ctx.counts {
if c > 0 {
return false
}
}
return true
}

r := idx / ctx.w
c := idx % ctx.w

for _, typeID := range ctx.types {
if ctx.counts[typeID] > 0 {
variations := ctx.anchoredShapes[typeID]
for _, shape := range variations {
fits := true
for _, pt := range shape.points {
pr, pc := r+pt.r, c+pt.c
if pr < 0 || pr >= ctx.h || pc < 0 || pc >= ctx.w {
fits = false
break
}
if grid[pr*ctx.w+pc] {
fits = false
break
}
}

if fits {
for _, pt := range shape.points {
grid[(r+pt.r)*ctx.w+(c+pt.c)] = true
}
ctx.counts[typeID]--

if backtrack(idx+1, grid, ctx, skipsLeft) {
return true
}

ctx.counts[typeID]++
for _, pt := range shape.points {
grid[(r+pt.r)*ctx.w+(c+pt.c)] = false
}
}
}
}
}

if skipsLeft > 0 {
grid[idx] = true
if backtrack(idx+1, grid, ctx, skipsLeft-1) {
return true
}
grid[idx] = false
}

return false
}

File: 12/gemini-3.0-pro/rust/src/main.rs
use std::collections::{HashMap, HashSet};
use std::env;
use std::fs;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct Point {
    r: i32,
    c: i32,
}

#[derive(Debug, Clone)]
struct RegionTask {
    w: i32,
    h: i32,
    counts: Vec<usize>,
}

#[derive(Debug, Clone)]
struct AnchoredShape {
    points: Vec<Point>,
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_file = if args.len() > 1 { &args[1] } else { "input.txt" };

    let content = fs::read_to_string(input_file).expect("Failed to read input file");
    let content = content.replace("\r\n", "\n");
    let parts: Vec<&str> = content.split("\n\n").collect();

    let mut shapes: HashMap<usize, Vec<Vec<Point>>> = HashMap::new();
    let mut regions: Vec<RegionTask> = Vec::new();

    for part in parts {
        let lines: Vec<&str> = part.trim().split('\n').collect();
        if lines.is_empty() {
            continue;
        }

        if lines[0].contains(':') {
            if lines[0].contains('x') { // Region
                for line in lines {
                    if !line.trim().is_empty() {
                        parse_region(line, &mut regions);
                    }
                }
            } else { // Shape
                parse_shape(&lines, &mut shapes);
            }
        }
    }

    let mut solved_count = 0;
    for region in regions {
        if solve_region(&region, &shapes) {
            solved_count += 1;
        }
    }

    println!("Day 12 Part 1: {}", solved_count);
    // println!("Day 12 Part 2: 0");
}

fn parse_shape(lines: &[&str], shapes: &mut HashMap<usize, Vec<Vec<Point>>>) {
    let header = lines[0].trim().trim_end_matches(':');
    let id: usize = header.parse().unwrap();

    let mut points: HashSet<Point> = HashSet::new();
    for (r, line) in lines.iter().enumerate().skip(1) {
        for (c, ch) in line.chars().enumerate() {
            if ch == '#' {
                points.insert(Point {
                    r: (r - 1) as i32,
                    c: c as i32,
                });
            }
        }
    }
    shapes.insert(id, generate_variations(&points));
}

fn parse_region(line: &str, regions: &mut Vec<RegionTask>) {
    let p: Vec<&str> = line.split(':').collect();
    let dims: Vec<&str> = p[0].split('x').collect();
    let w: i32 = dims[0].parse().unwrap();
    let h: i32 = dims[1].parse().unwrap();

    let count_strs: Vec<&str> = p[1].trim().split_whitespace().collect();
    let mut counts: Vec<usize> = Vec::new();
    for s in count_strs {
        counts.push(s.parse().unwrap());
    }
    regions.push(RegionTask { w, h, counts });
}

fn generate_variations(original: &HashSet<Point>) -> Vec<Vec<Point>> {
    let mut variations: Vec<Vec<Point>> = Vec::new();
    let mut seen: HashSet<String> = HashSet::new();

    let mut current = original.clone();
    for _ in 0..2 { // Flip
        for _ in 0..4 { // Rotate
            let normalized = normalize(&current);
            let key = points_to_string(&normalized);
            if !seen.contains(&key) {
                variations.push(normalized);
                seen.insert(key);
            }
            current = rotate(&current);
        }
        current = flip(original);
    }
    variations
}

fn rotate(points: &HashSet<Point>) -> HashSet<Point> {
    let mut new_points: HashSet<Point> = HashSet::new();
    for p in points {
        new_points.insert(Point { r: p.c, c: -p.r });
    }
    new_points
}

fn flip(points: &HashSet<Point>) -> HashSet<Point> {
    let mut new_points: HashSet<Point> = HashSet::new();
    for p in points {
        new_points.insert(Point { r: p.r, c: -p.c });
    }
    new_points
}

fn normalize(points: &HashSet<Point>) -> Vec<Point> {
    if points.is_empty() {
        return Vec::new();
    }
    let mut min_r = 1000000;
    let mut min_c = 1000000;
    for p in points {
        if p.r < min_r {
            min_r = p.r;
        }
        if p.c < min_c {
            min_c = p.c;
        }
    }
    let mut res: Vec<Point> = Vec::new();
    for p in points {
        res.push(Point {
            r: p.r - min_r,
            c: p.c - min_c,
        });
    }
    res.sort();
    res
}

fn points_to_string(points: &[Point]) -> String {
    let mut s = String::new();
    for p in points {
        s.push_str(&format!("{},{};", p.r, p.c));
    }
    s
}

struct StaticContext {
    w: i32,
    h: i32,
    types: Vec<usize>,
    anchored_shapes: HashMap<usize, Vec<AnchoredShape>>,
}

fn solve_region(region: &RegionTask, shapes: &HashMap<usize, Vec<Vec<Point>>>) -> bool {
    let mut presents: Vec<usize> = Vec::new();
    let mut total_area = 0;
    for (i, &count) in region.counts.iter().enumerate() {
        for _ in 0..count {
            presents.push(i);
            total_area += shapes[&i][0].len();
        }
    }

    if total_area > (region.w * region.h) as usize {
        return false;
    }

    let max_skips = (region.w * region.h) as usize - total_area;
    let mut grid = vec![false; (region.w * region.h) as usize];

    let mut present_types_set: HashSet<usize> = HashSet::new();
    for &p in &presents {
        present_types_set.insert(p);
    }
    let mut present_types: Vec<usize> = present_types_set.into_iter().collect();
    present_types.sort_by(|a, b| shapes[b][0].len().cmp(&shapes[a][0].len()));

    let mut counts_map: HashMap<usize, usize> = HashMap::new();
    for &p in &presents {
        *counts_map.entry(p).or_insert(0) += 1;
    }
    
    let max_id = *present_types.iter().max().unwrap_or(&0);
    let mut counts = vec![0; max_id + 1];
    for (k, v) in counts_map {
        counts[k] = v;
    }

    let mut anchored_shapes: HashMap<usize, Vec<AnchoredShape>> = HashMap::new();
    for &type_id in &present_types {
        let mut list: Vec<AnchoredShape> = Vec::new();
        for v in &shapes[&type_id] {
            let anchor = v[0];
            let mut points: Vec<Point> = Vec::new();
            for p in v {
                points.push(Point {
                    r: p.r - anchor.r,
                    c: p.c - anchor.c,
                });
            }
            list.push(AnchoredShape { points });
        }
        anchored_shapes.insert(type_id, list);
    }

    let static_ctx = StaticContext {
        w: region.w,
        h: region.h,
        types: present_types,
        anchored_shapes,
    };

    backtrack(0, &mut grid, &static_ctx, &mut counts, max_skips)
}

fn backtrack(idx: usize, grid: &mut [bool], static_ctx: &StaticContext, counts: &mut [usize], skips_left: usize) -> bool {
    let mut curr_idx = idx;
    while curr_idx < grid.len() && grid[curr_idx] {
        curr_idx += 1;
    }

    if curr_idx == grid.len() {
        for &c in counts.iter() {
            if c > 0 {
                return false;
            }
        }
        return true;
    }

    let r = (curr_idx as i32) / static_ctx.w;
    let c = (curr_idx as i32) % static_ctx.w;

    for &type_id in &static_ctx.types {
        if counts[type_id] > 0 {
            let variations = &static_ctx.anchored_shapes[&type_id];
            for shape in variations {
                let mut fits = true;
                for pt in &shape.points {
                    let pr = r + pt.r;
                    let pc = c + pt.c;
                    if pr < 0 || pr >= static_ctx.h || pc < 0 || pc >= static_ctx.w {
                        fits = false;
                        break;
                    }
                    if grid[(pr * static_ctx.w + pc) as usize] {
                        fits = false;
                        break;
                    }
                }

                if fits {
                    for pt in &shape.points {
                        grid[((r + pt.r) * static_ctx.w + (c + pt.c)) as usize] = true;
                    }
                    counts[type_id] -= 1;

                    if backtrack(curr_idx + 1, grid, static_ctx, counts, skips_left) {
                        return true;
                    }

                    counts[type_id] += 1;
                    for pt in &shape.points {
                        grid[((r + pt.r) * static_ctx.w + (c + pt.c)) as usize] = false;
                    }
                }
            }
        }
    }

    if skips_left > 0 {
        grid[curr_idx] = true;
        if backtrack(curr_idx + 1, grid, static_ctx, counts, skips_left - 1) {
            return true;
        }
        grid[curr_idx] = false;
    }

    false
}

